.github/workflows/test.yml

```diff
@@ -0,0 +1,227 @@
+          cat "$HOME/.bashrc"
+
+      - name: Apt-Get Install
+        if: startsWith(matrix.os, 'ubuntu')
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y shellcheck sqlite3 at
+
+      - name: MacOS DNS Patch
+        if: startsWith(matrix.os, 'macos')
+        run: |
+          # apply DNS patch
+          hostuserutil="$(python3 -c '
+          import cylc.flow.hostuserutil
+          print(cylc.flow.hostuserutil.__file__)
+          ')"
+          patch "${hostuserutil}" < rose/etc/conf/macos-patch
+
+      - name: Install Rose
+        working-directory: rose
+        run: |
+          pip install ."[all]"
+          pip install --no-deps git+https://github.com/cylc/cylc-rose.git
+          yarn install
+
+      - name: Checkout FCM
+        if: startsWith(matrix.os, 'ubuntu')
+        uses: actions/checkout@v2
+        with:
+          repository: ${{ github.event.inputs.fcm_repo || 'metomi/fcm' }}
+          ref: ${{ github.event.inputs.fcm_ref || 'master' }}
+          path: 'fcm'
+
+      - name: Install FCM
+        if: startsWith(matrix.os, 'ubuntu')
+        run: |
+          # install FCM deps
+          sudo apt-get install -y \
+            subversion \
+            build-essential \
+            gfortran \
+            libxml-parser-perl \
+            libconfig-inifiles-perl \
+            libdbi-perl \
+            libdbd-sqlite3-perl
+
+          # install wandisco
+          sudo sh -c 'echo "deb http://opensource.wandisco.com/ubuntu \
+            `lsb_release -cs` svn19" \
+            >> /etc/apt/sources.list.d/subversion19.list'
+          sudo wget -q http://opensource.wandisco.com/wandisco-debian.gpg -O- \
+            | sudo apt-key add -
+
+          # prepend FCM bin to $PATH
+          FCM_PATH="$GITHUB_WORKSPACE/fcm/bin"
+          # the github actions way (needed for cylc jobs)
+          echo "$FCM_PATH" >> "${GITHUB_PATH}"
+          # the bashrc wat (needed for subsequent gh action steps)
+          echo "export PATH=\"$FCM_PATH:\$PATH\"" >> "$HOME/.bashrc"
+
+      - name: Style
+        working-directory: rose
+        run: |
+          flake8
+          etc/bin/shellchecker
+          yarn run lint
+
+      - name: Unit Tests
+        working-directory: rose
+        run: |
+          pytest
+
+      - name: Functional Tests
+        timeout-minutes: 30
+        id: functest
+        working-directory: rose
+        env:
+          OBJC_DISABLE_INITIALIZE_FORK_SAFETY: YES
+        run: |
+          # rose tests should pass first time around
+          etc/bin/rose-test-battery -j 4 --state=save
+
+      - name: Re-Run Fails
+        if: failure() && steps.functest.outcome == 'failure'
+        working-directory: rose
+        env:
+          OBJC_DISABLE_INITIALIZE_FORK_SAFETY: YES
+        run: |
+          # so we only re-run for debug purposes
+          cylc scan --state=all --color=never
+          etc/bin/rose-test-battery -j 1 -v --state=save,failed
```

Comment:
So is the idea that if the tests fail the 1st time, the whole actions run should fail even if the tests somehow pass on retry?

Reply:
Yes.

Tests should be deterministic, testing shouldn't be a polling exercise of running and re running tests until they pass.

Unfortunately with Cylc we just don't have the time at our disposal to re-write the test battery so have had to live with the idea of accepting tests that pass on the second run but thankfully the Rose test battery is much more stable so doesn't require this treatment.